client/consensus/grandpa/authorities.go

```diff
@@ -0,0 +1,881 @@
+	var forced *hashNumber[H, N]
+	for _, c := range authSet.pendingForcedChanges {
+		isDesc, err := isDescendentOf(c.canonHash, bestHash)
+		if err != nil {
+			return nil, err
+		}
+		if !isDesc {
+			continue
+		}
+		forced = &hashNumber[H, N]{
+			hash:   c.canonHash,
+			number: c.canonHeight,
+		}
+		break
+	}
+
+	var standard *hashNumber[H, N]
+	for _, changeNode := range authSet.pendingStandardChanges.Roots() {
+		c := changeNode.change
+		isDesc, err := isDescendentOf(c.canonHash, bestHash)
+		if err != nil {
+			return nil, err
+		}
+		if !isDesc {
+			continue
+		}
+		standard = &hashNumber[H, N]{
+			hash:   c.canonHash,
+			number: c.canonHeight,
+		}
+		break
+	}
+
+	switch {
+	case standard != nil && forced != nil:
+		if forced.number < standard.number {
+			return forced, nil
+		}
+		return standard, nil
+	case forced != nil:
+		return forced, nil
+	case standard != nil:
+		return standard, nil
+	default:
+		return nil, nil
+	}
+}
+
+func (authSet *AuthoritySet[H, N]) addStandardChange(pending PendingChange[H, N], isDescendentOf IsDescendentOf[H]) error {
+	hash := pending.canonHash
+	number := pending.canonHeight
+
+	logger.Debugf(
+		"inserting potential standard set hashNumber signaled at block %d (delayed by %d blocks).",
+		number, pending.delay,
+	)
+
+	_, err := authSet.pendingStandardChanges.Import(hash, number, pending, isDescendentOf)
+	if err != nil {
+		return err
+	}
+
+	logger.Debugf(
+		"There are now %d alternatives for the next pending standard hashNumber (roots), "+
+			"and a total of %d pending standard changes (across all forks)",
+		len(authSet.pendingStandardChanges.Roots()), len(authSet.pendingStandardChanges.PendingChanges()),
+	)
+
+	return nil
+}
+
+func (authSet *AuthoritySet[H, N]) addForcedChange(pending PendingChange[H, N], isDescendentOf IsDescendentOf[H]) error {
+	for _, change := range authSet.pendingForcedChanges {
+		if change.canonHash == pending.canonHash {
+			return errDuplicateAuthoritySetChanges
+		}
+
+		isDescendent, err := isDescendentOf(change.canonHash, pending.canonHash)
+		if err != nil {
+			return fmt.Errorf("addForcedChange: checking isDescendentOf: %w", err)
+		}
+
+		if isDescendent {
+			return errMultiplePendingForcedAuthoritySetChanges
+		}
+	}
+
+	type key[N any] struct {
+		effectiveNumber   N
+		signalBlockNumber N
+	}
+
+	k := key[N]{
+		pending.EffectiveNumber(),
+		pending.canonHeight,
+	}
+
+	// Search by effective key
+	idx, _ := slices.BinarySearchFunc(
+		authSet.pendingForcedChanges,
+		k,
+		func(change PendingChange[H, N], k key[N]) int {
+			switch {
+			case change.EffectiveNumber() == k.effectiveNumber && change.canonHeight == k.signalBlockNumber:
+				return 0
+			case change.EffectiveNumber() > k.effectiveNumber && change.canonHeight > k.signalBlockNumber:
+				return 1
+			case change.EffectiveNumber() < k.effectiveNumber && change.canonHeight < k.signalBlockNumber:
+				return -1
+			default:
+				panic("huh?")
+			}
```

Comment:
I don't think having the panic in this binary search makes sense because you can cover all the comparison cases, here is my suggestion:

Implement the methods `LessThan` and `GreaterThan` in the `PendingChange` struct

```go
func (pc *PendingChange[H, N]) GreaterThan(effectiveBlockNumber, signalBlockNumber N) bool {
	effectiveNumberGreaterThan := pc.EffectiveNumber() > effectiveBlockNumber
	cannonHeighGreaterThan := pc.EffectiveNumber() == effectiveBlockNumber &&
		pc.canonHeight > signalBlockNumber

	return effectiveNumberGreaterThan || cannonHeighGreaterThan
}

func (pc *PendingChange[H, N]) LessThan(effectiveBlockNumber, signalBlockNumber N) bool {
	effectiveNumberLessThan := pc.EffectiveNumber() < effectiveBlockNumber
	cannonHeighLessThan := pc.EffectiveNumber() == effectiveBlockNumber &&
		pc.canonHeight < signalBlockNumber

	return effectiveNumberLessThan || cannonHeighLessThan
}
```

And in the `slices.BinarySearchFunc`:

```go
idx, _ := slices.BinarySearchFunc(
	authSet.pendingForcedChanges,
	k,
	func(change PendingChange[H, N], k key[N]) int {
		switch {
		case change.LessThan(k.effectiveNumber, k.signalBlockNumber):
			return -1
		case change.GreaterThan(k.effectiveNumber, k.signalBlockNumber):
			return 1
		default:
			return 0
		}
	},
)
```